### PR TITLE
[KYUUBI #377] build/dist support --spark-provided

### LINKED ISSUE #377: Make distribution bundle Spark optional
**Is your feature request related to a problem? Please describe.**
Currently, `build/dist` always bundle `spark-bin`, which not always required and significant increase the distribution bundle size.

**Describe the solution you'd like**
Add new an option `--spark-provided` on `build/dist` to make distribution bundle `spark-bin` optional.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by pan3793**
pan3793 [Closes #378](https://github.com/yaooqinn/kyuubi/pull/378)    [&#10088;?&#10089;](https://pullrequestbadge.com/?utm_medium=github&utm_source=yaooqinn&utm_campaign=badge_info)

### _Why are the changes needed?_

close #377

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

- [x] Manual test 

`build/dist --spark-provided`

```
(kyuubi) ➜  kyuubi git:(master) ✗ tree dist
dist
├── LICENSE
├── RELEASE
├── bin
│   ├── kyuubi
│   ├── kyuubi-logo
│   └── load-kyuubi-env.sh
├── conf
│   ├── kyuubi-defaults.conf.template
│   ├── kyuubi-env.sh.template
│   └── log4j.properties.template
├── externals
│   └── engines
│       └── spark
│           └── kyuubi-spark-sql-engine-1.1.0-SNAPSHOT.jar
├── jars
│   ├── apacheds-i18n-2.0.0-M15.jar
│   ├── apacheds-kerberos-codec-2.0.0-M15.jar
│   ├── api-asn1-api-1.0.0-M20.jar
│   ├── api-util-1.0.0-M20.jar
│   ├── audience-annotations-0.5.0.jar
│   ├── commons-beanutils-1.7.0.jar
│   ├── commons-beanutils-core-1.8.0.jar
│   ├── commons-codec-1.4.jar
│   ├── commons-collections-3.2.2.jar
│   ├── commons-configuration-1.6.jar
│   ├── commons-digester-1.8.jar
│   ├── commons-io-2.4.jar
│   ├── commons-lang-2.6.jar
│   ├── commons-lang3-3.10.jar
│   ├── commons-math-2.2.jar
│   ├── curator-client-2.7.1.jar
│   ├── curator-framework-2.7.1.jar
│   ├── curator-recipes-2.7.1.jar
│   ├── curator-test-2.7.1.jar
│   ├── guava-11.0.2.jar
│   ├── hadoop-annotations-2.7.4.jar
│   ├── hadoop-auth-2.7.4.jar
│   ├── hadoop-common-2.7.4.jar
│   ├── hive-service-rpc-2.3.7.jar
│   ├── httpclient-4.5.6.jar
│   ├── httpcore-4.4.12.jar
│   ├── javassist-3.18.1-GA.jar
│   ├── jcl-over-slf4j-1.7.30.jar
│   ├── jsr305-3.0.2.jar
│   ├── kyuubi-common-1.1.0-SNAPSHOT.jar
│   ├── kyuubi-ha-1.1.0-SNAPSHOT.jar
│   ├── kyuubi-main-1.1.0-SNAPSHOT.jar
│   ├── libfb303-0.9.3.jar
│   ├── libthrift-0.9.3.jar
│   ├── log4j-1.2.17.jar
│   ├── netty-3.7.0.Final.jar
│   ├── scala-library-2.12.10.jar
│   ├── slf4j-api-1.7.30.jar
│   ├── slf4j-log4j12-1.7.30.jar
│   ├── spotbugs-annotations-3.1.9.jar
│   └── zookeeper-3.4.14.jar
├── logs
├── pid
└── work
```

Bundle size
```
(kyuubi) ➜  kyuubi git:(KYUUBI-377) ll | grep tar.gz
-rw-r--r--   1 chengpan  staff   230M Feb 26 12:59 kyuubi-1.1.0-SNAPSHOT-bin-spark-3.0.2.tar.gz
-rw-r--r--   1 chengpan  staff    20M Feb 26 12:56 kyuubi-1.1.0-SNAPSHOT-bin-without-spark.tar.gz
```

### CODE REVIEW

**Comment by ulysses-you** [comment]
If specify `spark-provider`, it's no need to download spark bin file any more. Shall we skip the `kyuubi-download` module during build ?

**Comment by yaooqinn** [comment]
+1 for @ulysses-you's idea

**Comment by pan3793** [comment]
> If specify `spark-provider`, it's no need to download spark bin file any more. Shall we skip the `kyuubi-download` module during build ?

Make sense, I will change it.

**Comment by pan3793** [comment]
Updated.

File: build/dist [hunk 1]
```diff
@@ -136,7 +141,11 @@ HIVE_VERSION=$("$MVN" help:evaluate -Dexpression=hive.version $@ 2>/dev/null\
 echo "Building Kyuubi package of version $VERSION against Spark version - $SPARK_VERSION"
 
 if [[ "$NAME" == "none" ]]; then
-  NAME="spark-"$SPARK_VERSION
+  if [[ "$SPARK_PROVIDED" == "true" ]]; then
+    NAME="without-spark"
+  else
+    NAME="spark-"$SPARK_VERSION
```
Comment (yaooqinn) [inline build/dist:147]: Can we change it to use `major.minor` version form?`, also cc @ulysses-you WDYT?

Comment (pan3793) [reply build/dist:147]: I'm ok with this change, but it better change in separate PR.

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
+1. LGTM! 

TODOS:
1. use `major.minor` version for releases with spark
2. official releases automatically deploy a spark-provided one?

**Comment by ulysses-you** [comment]
thanks merging to master for 1.1

**Pull request closed without merging**